boot/BootFromDevice.c

```diff
@@ -135,47 +143,78 @@ int BootFromNative(CONFIGENTRY *config) {
 	return result;
 }
 
-CONFIGENTRY *DetectSystemFatX(void) {
+CONFIGENTRY *DetectSystemFatX() {
+	FATXPartitionTable *partitionTable;
 	CONFIGENTRY *config = NULL;
 	CONFIGENTRY *cfgLinux;
 	CONFIGENTRY *cfgReactOS;
-	FATXPartition *partition;
+	char entryName[MAX_CONFIG_TITLE];
+	int partIdx;
+
+	for (int driveId = 0; driveId < 1; driveId++) {
+		// Check if it's a DVD/CD drive
+		if (!tsaHarddiskInfo[driveId].m_fDriveExists) {
+			printk("DetectSystemFatX: Drive %d doesn't exist\n", driveId);
+			continue;
+		}
+		else if (tsaHarddiskInfo[driveId].m_fAtapi) {
+			#ifdef FATX_DEBUG
+			printk("DetectSystemFatX: Drive %d is ATAPI\n", driveId);
+			#endif
+			continue;
+		}
 
-	partition = OpenFATXPartition(0, SECTOR_STORE, STORE_SIZE);
-	if (!partition)
-		return NULL;
+		partitionTable = OpenFatXPartitionTable(driveId);
+		if (partitionTable == NULL) {
+			continue;
+		}
 
-	cfgLinux = DetectLinuxFATX(partition);
-	if (cfgLinux != NULL) {
-		FillConfigEntries(cfgLinux, BOOT_FATX, 0, 0);
-		config = AddNestedConfigEntry(config, cfgLinux, "Linux");
-	}
-	cfgReactOS = DetectReactOSFATX(partition);
-	if (cfgReactOS != NULL) {
-		FillConfigEntries(cfgReactOS, BOOT_FATX, 0, 0);
-		config = AddNestedConfigEntry(config, cfgReactOS, "ReactOS");
-	}
-	CloseFATXPartition(partition);
+		for (int sortPartIdx = 0; sortPartIdx < FATX_XBPARTITIONER_PARTITIONS_MAX; sortPartIdx++) {
+			partIdx = FATX_PARTITION_SORT_ORDER[sortPartIdx];
+			if (partitionTable->partitions[partIdx] == NULL) {
+				continue;
+			}
+
+			cfgLinux = DetectLinuxFATX(partitionTable->partitions[partIdx]);
+			if (cfgLinux != NULL) {
+				// TODO: Indicate non-default HDD here once that's tested and working
+				sprintf(entryName, "Linux (%c:)", FATX_DRIVE_LETTERS[partIdx]);
+				FillConfigEntries(cfgLinux, BOOT_FATX, driveId, partIdx);
+				config = AddNestedConfigEntry(config, cfgLinux, entryName);
+			}
 
+			cfgReactOS = DetectReactOSFATX(partitionTable->partitions[partIdx]);
+			if (cfgReactOS != NULL) {
+				FillConfigEntries(cfgReactOS, BOOT_FATX, driveId, partIdx);
+				// TODO: Indicate non-default HDD here once that's tested and working
+				sprintf("ReactOS (%c:)", FATX_DRIVE_LETTERS[partIdx]);
```

Comment:
Missing `entryName` as first argument here.
```suggestion
				sprintf(entryName, "ReactOS (%c:)", FATX_DRIVE_LETTERS[partIdx]);
```